File: csharp/need-for-speed/NeedForSpeed.cs
using System;

class RemoteControlCar
{
    private readonly int speed;
    private readonly int batteryDrain;

    private int distanceDriven = 0;
    private int batteryRemaining = 100;

    // TODO: define the constructor for the 'RemoteControlCar' class
    public RemoteControlCar(int speed, int batteryDrain)
    {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
    }

    public bool BatteryDrained()
    {
        return this.batteryRemaining < this.batteryDrain;
    }

    public int DistanceDriven()
    {
        return this.distanceDriven;
    }

    public void Drive()
    {
        if (!this.BatteryDrained())
        {
            this.batteryRemaining -= this.batteryDrain;
            this.distanceDriven += speed;
        }
    }

    public static RemoteControlCar Nitro()
    {
        return new RemoteControlCar(50, 4);
    }
}

class RaceTrack
{
    private readonly int distance;

    public RaceTrack(int distance)
    {
        this.distance = distance;
    }

    public bool TryFinishTrack(RemoteControlCar car)
    {
        do
        {
            car.Drive();
        } while (car.DistanceDriven() < this.distance && !car.BatteryDrained());

        return car.DistanceDriven() >= this.distance;
    }
}

File: java/need-for-speed/src/main/java/NeedForSpeed.java
class NeedForSpeed {
    // TODO: define the constructor for the 'NeedForSpeed' class

    private int batteryDrain;
    private int speed;
    private int batteryCharge = 100;
    private int distanceDriven = 0;

    public NeedForSpeed(int speed, int batteryDrain) {
        this.batteryDrain = batteryDrain;
        this.speed = speed;
    }

    public boolean batteryDrained() {
        return this.batteryCharge <= 0;
    }

    public int distanceDriven() {
        return this.distanceDriven;
    }

    public void drive() {
        this.batteryCharge -= this.batteryDrain;
        if (this.batteryCharge >= 0) {
            this.distanceDriven += this.speed;
        }
    }

    public static NeedForSpeed nitro() {
        return new NeedForSpeed(50, 4);
    }
}

class RaceTrack {
    private int distance;

    public RaceTrack(int distance) {
        this.distance = distance;
    }

    public boolean tryFinishTrack(NeedForSpeed car) {
        while (car.distanceDriven() < this.distance && !car.batteryDrained()) {
            car.drive();
        }

        return car.distanceDriven() >= this.distance;
    }
}
